### DISCUSSION TITLE
Change search path for dependency()?

### DISCUSSION THREAD

**Discussion by DidierMalenfant** [mesonbuild/meson#12070] [Q&A] [answered]
I'm trying to build glib, which uses mason as its build system, as a universal binary on macOS.

When linking it resolves `libpcre` to my system one (which is not a universal binary) instead of the one I've built myself (which is). The linker command line contains:
```
-Wl,-rpath,/opt/homebrew/Cellar/pcre2/10.42/lib /opt/homebrew/Cellar/pcre2/10.42/lib/libpcre2-8.dylib
```
Now this seems to be the result of:
```
      pcre2 = cc.find_library('pcre2-8', required : false)
```
in the `mason.build` file but I can't find a way to affect where this `find_library` is looking for its target.

I've tried setting `LIBRARY_PATH` in my env, tried passing `-L` options to the linker via `c_link_args` to no avail.

Ideally I'm looking for an environment or command line solution as I'm trying not to modify the library's files if possible.

**Comment by eli-schwartz** [comment, marked as answer]
> Now this seems to be the result of:
> 
> ```
>       pcre2 = cc.find_library('pcre2-8', required : false)
> ```
> 
> in the `mason.build` file but I can't find a way to affect where this `find_library` is looking for its target.

That's only used on Windows. The full check is:
```meson
pcre2 = dependency('libpcre2-8', version: pcre2_req, required: false, default_options: ['default_library=static'])
if not pcre2.found()
  if cc.get_id() == 'msvc' or cc.get_id() == 'clang-cl'
  # MSVC: Search for the PCRE2 library by the configuration, which corresponds
  # to the output of CMake builds of PCRE2.  Note that debugoptimized
  # is really a Release build with .PDB files.
    if vs_crt == 'debug'
      pcre2 = cc.find_library('pcre2d-8', required : false)
    else
      pcre2 = cc.find_library('pcre2-8', required : false)
    endif
  endif
endif

# Try again with the fallback
if not pcre2.found()
  pcre2 = dependency('libpcre2-8', version: pcre2_req, allow_fallback: true, default_options: ['default_library=static'])
  assert(pcre2.type_name() == 'internal')

...
```

Notice:
- dependency() is always used first, and checks with `pkg-config`, so you can override this by setting $PKG_CONFIG_PATH
- when using msvc `cl.exe` or `clang-cl.exe`, and pkg-config failed, try searching the `cc.find_library()` search path
- final fallback is to retry dependency() knowing it will fail, in order to use the wrap dependency fallback mechanism

**Reply by DidierMalenfant** [reply]
Yep, I stand corrected, the `dependency` call is indeed where the magic takes place. Thanks!

So I'm assuming just setting `PKG_CONFIG_PATH` before calling `meson setup` will do the right thing?

Is there a way to list all the dependency paths found after a `setup` command kinda like `cmake` does?

**Reply by DidierMalenfant** [reply]
Ok so it looks like `PKG_CONFIG_PATH` is not a path for libraries but a path for `pkg-config` config files. Since my build of `pcre2` doesn't seem to generate those, I manually created one as a test. This now works as expected:
```
PKG_CONFIG_PATH=/Users/didier/Desktop/FSUAE-Build/deps/lib/pkgconfig pkg-config --libs pcre2-8
-L/Users/didier/Desktop/FSUAE-Build/deps/lib -lpcre2-8
```
same path can then be passed to configure the mason build:
```
PKG_CONFIG_PATH=/Users/didier/Desktop/FSUAE-Build/deps/lib/pkgconfig meson setup build ...
```

**Reply by eli-schwartz** [reply]
If you look at the log file in builddir/meson-logs/meson-log.txt it will produce a debug report of the exact pkg-config command line run, with the environment variables it uses, and the stdout/stderr that the subprocess produced.

**Reply by DidierMalenfant** [reply]
Thank you for your help!
